### PR TITLE
[nnx] add performance guide notebook

### PR DESCRIPTION
**Pull request by cgarciae**
# What does this PR do?

* Add the Performance Considerations guide
* Updates the Flax Basics guide with a note of `nnx.jit`. Also reverts some formatting style changes made recently.
* Adds `treescope` as a direct dependency of Flax.

### CODE REVIEW

File: docs_nnx/guides/performance.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+  metrics.update(loss=loss)
+
+  return loss
+  
+for _ in range(10):
+  x, y = jnp.ones((32, 2)), jnp.zeros((32, 3))
+  loss = train_step(model, optimizer, metrics, x, y)
+```
+
+To speed it up we can use `nnx.split` before starting the training loop to create a `graphdef` and `state` pytrees for the Flax NNX objects as a group since `graphdef` and `state` are fast to traverse. Then at the beggining and end of a `jax.jit`-decorated function we can call `nnx.merge` and `nnx.split` to switch back and forth between the object and pytree representations. The important thing here is that `split` and `merge` will only run once during tracing.
```
Comment (IvyZX) [inline docs_nnx/guides/performance.md:72]: beggining -> beginning

Comment (IvyZX) [inline docs_nnx/guides/performance.md:72]: I think we should add a line here to explicitly say that you need to run nnx.merge and nnx.split on all the NNX objects that are part of the train step inputs.

**Review by IvyZX — APPROVED** [review]

File: docs_nnx/nnx_basics.md [hunk 2]
```diff
@@ -43,17 +28,9 @@ import jax.numpy as jnp
 
 ## The Flax NNX Module system
 
-The main difference between the Flax[`nnx.Module`](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/module.html) and other `Module` systems in [Flax Linen](https://flax-linen.readthedocs.io/en/latest/api_reference/flax.linen/module.html) or [Haiku](https://dm-haiku.readthedocs.io/en/latest/notebooks/basics.html#Built-in-Haiku-nets-and-nested-modules) is that in NNX everything is **explicit**. This  means, among other things, that:
-
-1) The [`nnx.Module`](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/module.html) itself holds the state (such as parameters) directly.
```
[the remark this answers is not in the document]
Comment (cgarciae) [reply docs_nnx/nnx_basics.md:48]: I've converted them to a single paragraph. I don't like excessive use of bullet points.

File: docs_nnx/nnx_basics.md [hunk 3]
```diff
@@ -12,27 +12,12 @@ jupytext:
 
 Flax NNX is a new simplified API that is designed to make it easier to create, inspect, debug, and analyze neural networks in [JAX](https://jax.readthedocs.io/). It achieves this by adding first class support for Python reference semantics. This allows users to express their models using regular Python objects, which are modeled as PyGraphs (instead of pytrees), enabling reference sharing and mutability. Such API design should make PyTorch or Keras users feel at home.
 
-In this guide you will learn about:
```
[the remark this answers is not in the document]
Comment (8bitmp3) [reply docs_nnx/nnx_basics.md:15]: This guide is an introduction to a lot of new concepts, I will add this back to help new users.

File: docs_nnx/nnx_basics.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (8bitmp3) [reply docs_nnx/nnx_basics.md:48]: This is a Flax vs Haiku comparison, so for ease of reading and to aid "scannability", I recommend using a list (without sequential numbering). A wall of text here may be missed by some users.

https://developers.google.com/style/accessibility#ease-of-reading

**Pull request merged**